tooltool.py

```diff
@@ -727,105 +730,153 @@ def purge(folder, gigs):
+    port = int(port)
+    conn = cls(host, port)
+    try:
+        req_path = "%s?%s" % (url.path, url.query) if url.query else url.path
+        conn.request('PUT', req_path, open(filename),
+                     {'Content-type': 'application/octet-stream'})
+        resp = conn.getresponse()
+        resp_body = resp.read()
+        conn.close()
+        if resp.status != 200:
+            raise RuntimeError("Non-200 return from AWS: %s %s\n%s" %
+                               (resp.status, resp.reason, resp_body))
+    except Exception:
+        file['upload_exception'] = sys.exc_info()
+        file['upload_ok'] = False
+    else:
+        file['upload_ok'] = True
 
-          * Do you agree to disclose your email address in association with this upload?
 
-        If you answered any of these questions with "no", please stop the
-        upload now.  Otherwise, hit enter to continue.
-        """)
-    raw_input()
-    log.info(
-        "The following three rsync commands will be executed to transfer the tooltool package:")
-    log.info("1) %s" % cmd1)
-    log.info("2) %s" % cmd2)
-    log.info("3) %s" % cmd3)
-    log.info("Please note that the order of execution IS relevant!")
-    log.info("Uploading hashed files with command: %s" % cmd1)
-    execute(cmd1)
-    log.info("Uploading metadata files (notes)    with command: %s" % cmd2)
-    execute(cmd2)
-    log.info("Uploading metadata files (manifest) with command: %s" % cmd3)
-    execute(cmd3)
+def _notify_upload_complete(base_url, auth_file, file):
+    req = urllib2.Request(
+        urlparse.urljoin(
+            base_url,
+            'upload/complete/%(algorithm)s/%(digest)s' % file))
+    _authorize(req, auth_file)
+    try:
+        urllib2.urlopen(req)
+    except urllib2.HTTPError as e:
+        if e.code != 409:
+            _log_api_error(e)
+            return
+        # 409 indicates that the upload URL hasn't expired yet and we
+        # should retry after a delay
+        to_wait = int(e.headers.get('X-Retry-After', 60))
+        log.warning("Waiting %d seconds for upload URLs to expire" % to_wait)
+        time.sleep(to_wait)
+        _notify_upload_complete(base_url, auth_file, file)
+    except Exception:
+        log.exception("While notifying server of upload completion:")
+
+
+def upload(manifest, message, base_urls, auth_file, region):
+    try:
+        manifest = open_manifest(manifest)
+    except InvalidManifest:
+        log.exception("failed to load manifest file at '%s'")
+        return False
 
-    log.info("Package %s has been correctly uploaded to %s:%s" %
-             (package, host, path))
+    # verify the manifest, since we'll need the files present to upload
+    if not manifest.validate():
+        log.error('manifest is invalid')
+        return False
 
-    return True
+    if any(fr.visibility is None for fr in manifest.file_records):
+        log.error('All files in a manifest for upload must have a visibility set')
```

Comment:
I have a vague memory, that we still run tooltool under python 2.6 in some cases (default python on CentOS 6). In this case any() will be failing.